FIPS: Fix compiler errors in rsa_chk.c when building with `-DFIPS_MODE`

mspncp opened the pull request:
~~This commit contains two fixes:~~

**Fix compiler error in rsa_chk.c**


This file currently does not compile with `-DFIPS_MODE`.

```make
~/src/openssl$ CFLAGS=-DFIPS_MODE perl ./Configure debug-linux-x86_64 enable-crypto-mdebug shared --strict-warnings --prefix=/opt/openssl-dev
~/src/openssl$ make

...

crypto/rsa/rsa_chk.c: In function 'RSA_check_key_ex':
crypto/rsa/rsa_chk.c:30:73: error: expected ')' before 'return'
             && rsa_sp800_56b_check_keypair(key, NULL, -1, RSA_bits(key))
                                                                         ^
                                                                         )
         return 0;
         ~~~~~~                                                           
crypto/rsa/rsa_chk.c:28:10: note: to match this '('
     if (!(rsa_sp800_56b_check_public(key)
          ^
crypto/rsa/rsa_chk.c:31:18: error: expected ')' before '}' token
         return 0;
                  ^
                  )
crypto/rsa/rsa_chk.c:242:1:
 }
 ~                 
crypto/rsa/rsa_chk.c:28:8: note: to match this '('
     if (!(rsa_sp800_56b_check_public(key)
        ^
crypto/rsa/rsa_chk.c:242:1: error: expected expression before '}' token
 }
 ^
crypto/rsa/rsa_chk.c:242:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1: all warnings being treated as errors
make[1]: *** [Makefile:10742: crypto/rsa/libcrypto-lib-rsa_chk.o] Error 1
make[1]: Leaving directory '/home/msp/openssl/openssl'
make: *** [Makefile:165: all] Error 2
```

_Note: To my surprise it seems like it is broken since  July 5, 2018 (commit 8240d5fa653). But maybe it's because I configured it in a non-standard fashion by adding the `-DFIPS_MODE` flag explicitly?_

_I did this in order to take a look at the CRNG test (#8790), because I was unsure whether the FIPS provider implementation is already in a usable state and how it is configured properly._


**~~Temporarily disable bignum tests in FIPS mode~~**


~~This is only a temporary workaround until the todo from 9efa0ae0b60 is fixed.~~

~~To make the change more easily revertible, I used a separate `#ifndef FIPS_DISABLE_BIGNUM_TESTS` instead of `#ifndef FIPS_MODE`.~~

mspncp commented:
> To make the change more easily revertible, I used a separate #ifndef FIPS_DISABLE_BIGNUM_TESTS instead of #ifndef FIPS_MODE.

Thinking once more about it, I think it's even better to have two separate commits. Hang on a minute...

paulidale commented:
These have been known for a while.  I created a PR that added a Travis build that included -DFIPS_MODE but it was rejected: #8659

mspncp commented:
I am aware that setting `-DFIPS_MODE` explicitly is a hack. But independently of that, shouldn't the compiler errors be fixed anyway? At least fix 2b4a470a65283bfe2336003d6eb7f6b229f30588 should be justified. The temporary fix for the tests 65c2dcdb73b2ffdc21c49d90c5eaafc77fe838aa might be arguable.

mspncp commented:
Returning to my original problem: Since the temporary `-DFIPS_MODE` hack was rejected in #8659,  is there any officially supported alternative way to end up with a binary where I can live debug the RNG test (#8790)?  For example, by running

```
util/shlib_wrap.sh gdb --args apps/openssl rand -hex 10
```

paulidale commented:
I add a `#define` to the file in question.

mspncp commented:
> I am aware that setting `-DFIPS_MODE` explicitly is a hack. But independently of that, shouldn't the compiler errors be fixed anyway? At least fix 2b4a470 should be justified. The temporary fix for the tests 65c2dcd might be arguable.

@mattcaswell what's your opinion? I can drop 65c2dcd  If you think it is unnecessary, but obviously 2b4a470 needs to be fixed.

mattcaswell commented:
I'm not keen on the 65c2dcd change. I'm obviously ok with the other one.

mspncp commented:
> I'm not keen on the 65c2dcd change. I'm obviously ok with the other one.

I almost anticipated you would say that. So I dropped 65c2dcd and adjusted the pr title accordingly.

mspncp commented:
Merged, thanks!

96384e613a FIPS: Fix compiler errors in rsa_chk.c when building with `-DFIPS_MODE`
